Add dynamic.layout chapter (#15)
rust-lang/spec: spec/dynamic/layout.md

[^6]: `repr(Rust)` types have an unspecified and unstable layout

[^7]: Same as equivalent tuple struct. Struct size and alignment requirement may be modified by the `#[repr(align)]`

[^8]: Same as equivalent unit struct

## Layout Properties [§](r[dynamic.layout.properties])

[§](r[dynamic.layout.properties.size]) Each `Sized` type has a property called it's size. The size is a non-negative integer.

[!NOTE]: For example, the size of `u8` is `1`, and the size of `u32` is `4`.

[§](r[dynamic.layout.properties.align]) Each `Sized` type and the slice type has a property call its alignment requirement.  The alignment requirement is a positive integer power of 2.

[!NOTE]: The alignment requirement may simply be called the alignment or align of the type.

[§](r[dynamic.layout.properties.size-align-value]) The size and alignment of any type is representible as a value of type `usize`. The alignment of a type divides its size.

[!NOTE]: The minimum size of a type is `0`, and the minimum alignment requirement is `1`. A type that has size `0` and alignment requirement `1` is called a 1-ZST.

> [§](r[dynamic.layout.properties.storage]) Storage for a value is suitable for storing a value of a `Sized` type if it is at least the size of that type and the address of the start of the storage satisfies the alignment requirement.

Reviewer: You forgot to define what it means to "satisfy the alignment requirement", the spec only ever said it was a pow2 number

[!NOTE]: Scalar types and pointer types have no padding bytes.

[§](r[dynamic.layout.properties.validity]) Each `Sized` type has a validity invariant, which constrains the values that can be read from storage. Only value bytes are taken into account in determining the validity of a value.

[§](r[dynamic.layout.properties.underlying]) Each `Sized` type may have an underlying type, which is also `Sized`. Such a type has the same size, alignmentment, representation, and pointer metadata type as the underlying type. Such a type may have an additional validity invariant, and also has the validity invariant of the underlying type.

[!NOTE]: A repr(transparent) type has its transparent field as an underlying type

[§](r[dynamic.layout.properties.ptr-metadata]) Each type has a property called the pointer metadata type. The pointer metadata type is used in determining the layout of pointers to that type.

[§](r[dynamic.layout.properties.ptr-metadata-thin]) The pointer metadata type of any `Thin` type is `()`.

## Scalar Layout [§](r[dynamic.layout.scalar])

[§](r[dynamic.layout.scalar.int-size]) Each integer type of width `N` has a size exactly equal to `N/8`. 

> [1NOTE]: The width of the integer type `uN` or `iN` is `N`. 

Reviewer: typo `!NOTE`

[§](r[dynamic.layout.pointer.fn-repr]) The address of a function pointer value is the same as the value of type `usize` computed from the same bytes. The pointer tag of a pointer value is each pointer portion of each byte of the representation. 

[§](r[dynamic.layout.pointer.fn-validity]) A value of an function pointer type is valid if the address of the pointer is not `0`.

## Aggregate Layout [§](r[dynamic.layout.aggregate])

[§](r[dynamic.layout.aggregate.fields]) Each aggregate type has a list of fields, that each have a type and offset. 

[!NOTE]: This includes built-in tuple types and arrays. This does not include `enum` definitions.

[§](r[dynamic.layout.aggregate.struct-base] The fields of a `struct` definition are each layed at offsets that ensure that each field occupies nonoverlapping storage, which, unless modified by the `repr(packed)` attribute (r#[dynamic.layout.aggregate.packed])) is suitably aligned. The size of a `struct` definition is at least sufficient to store each field, and the alignment requirement of a `struct` definition, unless modified by the `repr(packed)` attribute, is suitable to align each field.

> [!NOTE]: The offsets of the fields, sizes, and alignment requirements of two different `struct` types, even with the same field types in the same order, may be different.

Reviewer: maybe say "same source order" here? Not sure if that's an existing concept

> even with the same field types in the same declaration order, may be different unless modified by the `repr(C)` attribute.

Author: Source order is probably fine since it's it's a note.

[!NOTE]: The offsets of 1-ZST fields of such a `struct` are not specified, but do not exceed the size of the underlying type.

[§](r[dynamic.layout.aggregate.repr-align]) A `struct` or `union` definition with a `repr(align(N))` repr-attribute, where `N` is an integer power of 2 up to a target-defined value, has an alignment requirement of at least `N`. The `repr(align(N))` repr-attribute may not be combined with any `repr(packed(N))` repr-attributes.

[§](r[dynamic.layout.aggregate.repr-packed]) A `struct` or `union` definition with a `repr(packed(N))` repr-attribute, where `N` is an integer power of 2 up to a target-defined value, sets the alignment requirement of each field to the smallest of the alignment requirement of the type and `N`. The `repr(packed(N))` repr-attribute may not be combined with any `repr(align(N))` repr-attributes.

> [!NOTE]: The repr-attribute `repr(packed(1))` may be written as `repr(packed)`

Reviewer: I think this needs to be normative. `dynamic.layout.aggregate.repr-packed-short-form`

[§](r[dynamic.layout.aggregate.struct-metadata]) The pointer metadata type of a `struct` type is the pointer metadata type of the last field of the struct in declaration order.

[!NOTE]: If the struct has no fields, then it is `Sized` and thus has a pointer metadata type of `()`.

[§](r[dynamic.layout.aggregate.str]) The `str` type has an underlying type of `[u8]`.

[§](r[dynamic.layout.aggregate.never]) The `!` type has an underlying type of `()`. A value of type `!` is not valid.

## Trait Object Type [§](r[dynamic.layout.dyn])

[§](r[dynamic.layout.dyn.metadata]) A trait object type `dyn Trait+Markers` has an unspecified pointer metadata type that has an underlying type of `*const VTable<Trait>` where the type `VTable<Trait>` is an unspecified `Thin` type. 

[!NOTE]: The pointer metadata type is called the vptr type.

[§](r[dynamic.layout.dyn.metadata-repr]) The representation of the vptr type is an unspecified address and unspecified pointer tag.

[§](r[dynamic.layout.dyn.metadata-validity]) The validity invariant of the vptr type is unspecified, such that the result of unsizing a type as the trait object `dyn Trait+Markers` is valid.

> [§](r[dynamic.layout.dyn.metadata-size-align]) The size and alignment requirement acessible from a value of the vptr type obtained as the result of unsizing a type as the trait object `dyn Trait+Markers` is valid are the size and alignment of that type.

Reviewer: delete "is valid"

[§](r[dynamic.layout.enum.size-align]) The size of an `enum` definition is at least sufficient to store its representation. The alignment of an `enum` definition is at least sufficient to satisfy the alignment requirement of all fields of all variants in its representation.

[!NOTE]: In the case of an `enum` with a repr-attribute other than `repr(Rust)` or `repr(align(N))`, the size and alignment are given by the appropriate *exposition only* `Repr` type.

[§](r[dynamic.layout.enum.repr-align]) If the `align(N)` repr-attribute is present on an `enum` definition, then the alignment requirement of the enum is at least `N`.

[§](r[dynamic.layout.enum.option]) The special type `Option<T>` is an enum type, such that if `T` is one of the following types or a type with an underlying type that is one of the following (recursively), it has the corresponding underlying type.
* `&T`: `*const T`
* `&mut T`: `*mut T`

> * fn-ptr type: An unspecified, exposition-only, type with the same size and alignment as `fn()` that is valid for any initialized value

Reviewer: I can't tell whether or not this accidentally restricts the value from carrying pointer parts. Add more words to clarify

* `core::num::NonZeroUN`: `uN`
* `core::num::NonZeroIN`: `iN`
* `core::ptr::NonNull<T>`: `*mut T`
* `alloc::boxed::Box<T>`: `*mut T`

[§](r[dynamic.layout.enum.option-repr]) The representation of the value `None` of type `Option<T>` where `T` is any type refered to in r#[dynamic.layout.enum.option] is the initialized value with each value byte set to `0` with no pointer part, except that the metadata field of `Option<&T>`, `Option<&mut T>`, `Option<NonNull<T>>`, or `Option<Box<T>>` (where `T` is `!Thin`) is unspecified. The representation of the value `Some(x)` of type `Option<T>` where `T` is any type refered to in r#[dynamic.layout.enum.option] is the representation of `x`.

[§](r[dynamic.layout.enum.option-ref-validity]) A value of type `Option<&T>` or `Option<&mut T>` is valid if it corresponds to a pointer with address `0`, or a pointer with an address that satisfies the dynamic alignment requirement of `T`.

> [§](r[dynamic.layout.enum.result]) The special type `Result<T,E>` where `E` has size 0 and alignment 1, and `T` is a type mentioned in r#[dynamic.layout.enum.option], then `Result<T,E>` has an underlying type of `Option<T>`. The special type `Result<T,E>` where `T` has size 0 and alignment 1, and `E` is a type mentioned in r#[dynamic.layout.enum.option], then `Result<T,E>` has an underlying type of `Option<T>`

Reviewer: Final type mentioned should be `Option<E>`

> [§](r[dynamic.layout.enum.result-repr]) Where `Result<T,E>` is a type referred to by r#[dynamic.layout.enum.result] The representation of the value `Err(e)` of type `Result<T,E>` is the same as the representation for `Some(e)` of type `Option<E>` if `T` has size 0 and alignment 1, and `None` of type `Option<T>` otherwise, and the representation of `Ok(t)` of type `Result<T,E>` is the same as the representation for `Some(t)` of type `Option<T>` if `E` has size 0 and alignment 1, and `None` of the type `Option<E>` otherwise.

Reviewer: Please describe all the restrictions before the requirements, and also make `Option<T>` the first in the paragraph and `Option<E>` the second.

[§](r[dynamic.layout.properties.size]) Each `Sized` type has a property called it's size. The size is a non-negative integer.

[!NOTE]: For example, the size of `u8` is `1`, and the size of `u32` is `4`.

[§](r[dynamic.layout.properties.align]) Each `Sized` type and the slice type has a property called its alignment requirement.  The alignment requirement is a positive integer power of 2. An address satisfies an alignment requirement if it is a multiple of that alignment requirement. 

[!NOTE]: The alignment requirement may simply be called the alignment or align of the type.

[§](r[dynamic.layout.properties.size-align-value]) The size and alignment of any type is representible as a value of type `usize`. The alignment of a type divides its size.

[!NOTE]: The minimum size of a type is `0`, and the minimum alignment requirement is `1`. A type that has size `0` and alignment requirement `1` is called a 1-ZST.

[§](r[dynamic.layout.properties.storage]) Storage for a value is suitable for storing a value of a `Sized` type if it is at least the size of that type and the address of the start of the storage satisfies the alignment requirement.

> [§](r[dynamic.layout.properties.representation]) Each `Sized` type has a representation which determines how particular values of that type are layed out when read from or written to suitable storage. The representation of a type is exactly its size.

Reviewer 2: > The representation of a type is exactly its size.

I don't know what this means, a representation is a lot more than just the size?

From this definition I can't even tell if you mean what t-opsem calls "layout" or what we call "representation". I think you need to be more clear about what the types are of the object that you are defining.

"representation", in the t-opsem meaning, is a *relation* between values and byte lists. It encodes a ton of things, not just layout aspects but also validity.

"layout" is much more first-order, it stores concrete data. For instance, the "layout" of a struct is given by the offsets of all its fields as well as the total size and alignment of the struct. I think this chapter should only do layout, not representation (and therefore also not validity). Something like:

Each `Sized` type has an associated *layout* that stores all the information required to lay out the type in memory.

Note: The layout is then used by the [representation] to determine how [values] are related to lists of bytes that represent the value in memory.

(I think unsized types may actually also have layout but I'm okay with deferring this to later, I haven't yet explored unsized types in MiniRust)

Author: This is trying to encode that the list of bytes that any value becomes is exactly as long as the size property of the type.

Author: This is trying to prescribe that the list of bytes that any value becomes is exactly as long as the size property of the type.

[§](r[dynamic.layout.properties.padding]) The representation of a type may include padding bytes that do not participate in the computation of the value. All other bytes of the representation are called value bytes.

[!NOTE]: Scalar types and pointer types have no padding bytes.

> [§](r[dynamic.layout.properties.validity]) Each `Sized` type has a validity invariant, which constrains the values that can be read from storage. Only value bytes are taken into account in determining the validity of a value.

Reviewer 2: This is not following t-opsem terminology and also not following the structure that we've been exploring with MiniRust. I don't think the spec should be structured like this.

See [here](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#representation-relation) and [here](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#layout) for some t-opsem definitions, and [here](https://rust-lang.zulipchat.com/#narrow/stream/399173-t-spec/topic/representation.20and.20validity) for further Zulip discussion on this concern.

> [§](r[dynamic.layout.properties.underlying]) Each `Sized` type may have an underlying type, which is also `Sized`. Such a type has the same size, alignmentment, representation, and pointer metadata type as the underlying type. Such a type may have an additional validity invariant, and also has the validity invariant of the underlying type.

Reviewer 3: ```suggestion
[§](r[dynamic.layout.properties.underlying]) Each `Sized` type may have an underlying type, which is also `Sized`. Such a type has the same size, alignment, representation, and pointer metadata type as the underlying type. Such a type may have an additional validity invariant, and also has the validity invariant of the underlying type.
```

[!NOTE]: A repr(transparent) type has its transparent field as an underlying type

[§](r[dynamic.layout.properties.ptr-metadata]) Each type has a property called the pointer metadata type. The pointer metadata type is used in determining the layout of pointers to that type.

[§](r[dynamic.layout.properties.ptr-metadata-thin]) The pointer metadata type of any `Thin` type is `()`.

## Scalar Layout [§](r[dynamic.layout.scalar])

[§](r[dynamic.layout.scalar.int-size]) Each integer type of width `N` has a size exactly equal to `N/8`. 

[!NOTE]: The width of the integer type `uN` or `iN` is `N`. 

[§](r[dynamic.layout.scalar.int-align]) Each integer type has a target-dependendant alignment requirement which is at most its size. 

> [§](r[dynamic.layout.scalar.unsigned-repr]) Each unsigned integer type is represented as a sequence of bytes, in a target-dependenant order of increasing signifiance. The value, decomposed as groups of order 2^8, is layed out in the target-dependenant signifiance order.

Reviewer 3: ```suggestion
[§](r[dynamic.layout.scalar.unsigned-repr]) Each unsigned integer type is represented as a sequence of bytes, in a target-dependent order of increasing significance. The value, decomposed as groups of order 2^8, is laid out in the target-dependenant signifiance order.
```

[§](r[dynamic.layout.scalar.int-ptr-bytes]) Any pointer portion of a byte used in computing a value of an integer type is discarded.

[§](r[dynamic.layout.scalar.int-size]) The special types `usize` and `isize` are integer types with the same target-dependent width.

[!NOTE]: The width of `usize` and `isize` do not need to correspond to the width of any other integer type. If it does, then they have the same representation as that integer type.

[§](r[dynamic.layout.scalar.char]) The type `char` has an underlying type of `u32`. 

[§](r[dynamic.layout.scalar.char-validity]) A given value of type `char` is valid if the corresponding value of type `u32` is valid, and does not lie in the range `0xD800..0xE000` or the range `0x110000..`. 

[§](r[dynamic.layout.scalar.bool]) The type `bool` has an underlying type of `u8`.

[§](r[dynamic.layout.scalar.bool-repr]) The value `false` is represented the same as the value `0_u8`, and the value `true` is represented the same as the value `1_u8`

[§](r[dynamic.layout.scalar.bool-validity]) A given value of type `bool` is valid if the corresponding value of type `u8` is valid, and the value lies in the range `0..2`.

> [§](r[dynamic.layout.scalar.float]) A floating point type with width `N` has an underlying type of `uN`.

Reviewer 3: ```suggestion
[§](r[dynamic.layout.scalar.float]) A floating-point type with width `N` has an underlying type of `uN`.
```

> [§](r[dynamic.layout.scalar.float-repr]) A value of a floating-point has the representation corresponding to the value of the unsigned integer type computed from the value given by the appropriate interchange format from [IEEE 754](https://ieeexplore.ieee.org/document/5976968).

Reviewer 3: ```suggestion
[§](r[dynamic.layout.scalar.float-repr]) A floating-point value has the representation corresponding to the value of the unsigned integer type computed from the value given by the appropriate interchange format from [IEEE 754](https://ieeexplore.ieee.org/document/5976968).
```

[!NOTE]: The corresponding value may be called the bit representation of the floating-point value. 

## Pointer Layout [§](r[dynamic.layout.pointer])

[§](r[dynamic.layout.pointer.thin]) A pointer to a `Thin` type has a data portion only, which represents an address and a pointer tag. 

[!NOTE]: Such pointers are called thin pointers. All `Sized` types are also `Thin`. More types may be considered `Thin` in future versions of the spec. The layout of `*const T` and `*mut T` are the same.

[§](r[dynamic.layout.pointer.size-align]) The size and alignment of a thin pointer is the same as the size and alignment of `usize`. 

> [!NOTE] All thin pointers have the same

Reviewer 3: Seems to be missing something here.

> [§](r[dynamic.layout.pointer.repr]) The address of a pointer value is the same as the value of type `usize` computed from the same bytes. The pointer tag of a pointer value is each pointer portion of each byte of the representation. 

Reviewer 3: ```suggestion
[§](r[dynamic.layout.pointer.repr]) The target of a pointer value is the same as the value of type `usize` computed from the same bytes. The pointer tag of a pointer value is each pointer portion of each byte of the representation. 
```

Author: address is a term used elsewhere in rust. For example, under the strict provenance API, there are functions to extract the address part that use the name address.

Reviewer 3: > address is a term used elsewhere in rust. For example, under the strict provenance API, there are functions to extract the address part that use the name address.

Totally understand that. I was just trying to make some language to "show" the reader that the value (in this case) is what determines the target. 

Although, now that I reread it, I think that I understand the reason for the use of *address*: the reader will know (from other places in the standard) that the term identifies the thing being pointed to/at.

Sorry for the spam!

[!NOTE]: This may be called null pointer optimization or discriminant elision. It is guaranteed for the types mentioned above. 

[§](r[dynamic.layout.enum.option-ref-validity]) A value of type `Option<&T>` or `Option<&mut T>` is valid if it corresponds to a pointer with address `0`, or a pointer with an address that satisfies the dynamic alignment requirement of `T`.

[§](r[dynamic.layout.enum.result]) The special type `core::result::Result<T,E>` where `E` has size 0 and alignment 1, and `T` is a type mentioned in r#[dynamic.layout.enum.option], then `Result<T,E>` has an underlying type of `Option<T>`. The special type `Result<T,E>` where `T` has size 0 and alignment 1, and `E` is a type mentioned in r#[dynamic.layout.enum.option], then `Result<T,E>` has an underlying type of `Option<E>`

> [§](r[dynamic.layout.enum.result-repr]) Where `Result<T,E>` is a type referred to by r#[dynamic.layout.enum.result], if `E` is of size 0 and alignment 1, the representation of `Ok(t)` of type `Result<T,E>` is the same as the representation of `Some(t)` of type `Option<T>` and the represention of `Err(e)` is the same as the representation of `None` of type `Option<T>`. Otherwise, the representation of `Ok(t)` of type `Result<T,E>` is the same as the representation of `None` of type `Option<E>` and the representation of `Err(e)` of type `Result<T,E>` is the same as the representation of `Some(e)` of type `Option<E>`

Reviewer: Otherwise, where `T` is of size 0 and alignment 1,

Restating the condition for avoidance of editing mistakes, but using "where" instead of "if" because this is actually exhaustive.
